trie/hasher.go

```diff
@@ -177,30 +178,65 @@ func (h *hasher) hashChildren(original node, db DatabaseWriter) (node, node, err
 
 			// Reenable threading for subsequent hash calls
 			h.threaded = false
-		} else {
-			for i := 0; i < 16; i++ {
-				hashChild(i, nil)
-			}
+		} else {*/
+		for i := 0; i < 16; i++ {
+			hashChild(i, nil)
 		}
+		//}
 		if err != nil {
-			return original, original, err
+			return original, original, nil, err
 		}
 		cached.Children[16] = n.Children[16]
 		if collapsed.Children[16] == nil {
 			collapsed.Children[16] = valueNode(nil)
 		}
-		return collapsed, cached, nil
+		var refs []common.Hash
+		for i := 0; i < 16; i++ {
+			refs = append(refs, h.externals(collapsed.Children[i])...)
+		}
+		return collapsed, cached, refs, nil
 
 	default:
 		// Value and hash nodes don't have children so they're left as were
-		return n, original, nil
+		return n, original, h.externals(n), nil
+	}
+}
+
+// externals returns any external nodes referenced by a particular node. The only
+// current case for it is when an account trie references its storage trie.
+func (h *hasher) externals(n node) []common.Hash {
+	// Only value nodes can reference external nodes
+	val, ok := n.(valueNode)
+	if !ok {
+		return nil
+	}
+	// Account nodes have very specific sizes, discard anything else
+	// TODO(karalabe): Seriously? Dafuq man?!
+	if size := len(val); size < 70 || size > 102 {
```

Comment:
Why not pass in a function to decode values and return references to the trie constructor instead of this hack?